matrix/common/etl/__init__.py

```diff
@@ -0,0 +1,219 @@
+                             TableName.EXPRESSION.value),
+                exist_ok=True)
+
+    extractor = DSSExtractor(staging_directory=staging_directory,
+                             content_type_patterns=content_type_patterns,
+                             filename_patterns=filename_patterns,
+                             dss_client=get_dss_client(deployment_stage))
+    extractor.extract(
+        query=query,
+        transformer=transformer_cb,
+        finalizer=finalizer_cb,
+        max_workers=max_workers
+    )
+
+
+def transform_bundle(bundle_uuid: str, bundle_version: str, bundle_path: str, extractor: DSSExtractor):
+    """
+    Transforms a downloaded bundle into PSV rows - per bundle transformer passed to ETL.
+    :param bundle_uuid: Downloaded bundle UUID
+    :param bundle_version: Downloaded bundle version
+    :param bundle_path: Local path to downloaded bundle dir
+    :param extractor: ETL extractor object
+    """
+    logger.info(f"ETL: Downloaded bundle {bundle_uuid}.{bundle_version}. Transforming to PSV.")
+    transformers = [
+        CellExpressionTransformer(extractor.sd)
+    ]
+
+    for transformer in transformers:
+        transformer.transform(bundle_path)
+
+
+def finalizer_reload(extractor: DSSExtractor):
+    """
+    Final transformer during ETL, invokes loader - finalizer passed to ETL.
+    :param extractor: ETL extractor object
+    """
+    logger.info(f"ETL: All bundles downloaded. Performing final PSV transformations.")
+    job_id = str(uuid.uuid4())
+    transformers = [
+        FeatureTransformer(extractor.sd),
+        AnalysisTransformer(extractor.sd),
+        DonorLibraryTransformer(extractor.sd),
+        ProjectPublicationContributorTransformer(extractor.sd)
+    ]
+
+    for transformer in transformers:
+        transformer.transform(os.path.join(extractor.sd, 'bundles'))
+
+    logger.info(f"ETL: All transformations complete.")
+    _upload_to_s3(os.path.join(extractor.sd, MetadataToPsvTransformer.OUTPUT_DIRNAME), job_id)
+    _populate_all_tables(job_id, temp=False)
+
+
+def finalizer_update(extractor: DSSExtractor):
+    """
+    Final transformer during ETL for a single bundle
+    :param extractor:
+    :return:
+    """
+    logger.info(f"ETL: All bundles downloaded. Performing final PSV transformations.")
+    job_id = str(uuid.uuid4())
+    transformers = [
+        AnalysisTransformer(extractor.sd),
+        DonorLibraryTransformer(extractor.sd),
+        ProjectPublicationContributorTransformer(extractor.sd)
+    ]
+
+    for transformer in transformers:
+        transformer.transform(os.path.join(extractor.sd, 'bundles'))
+
+    logger.info(f"ETL: All transformations complete.")
+    _upload_to_s3(os.path.join(extractor.sd, MetadataToPsvTransformer.OUTPUT_DIRNAME), job_id)
+    _populate_all_tables(job_id, temp=True)
+
+
+def _upload_to_s3(output_dir, job_id: str):
+    """
+    Uploads to S3 all files in OUTPUT_DIR.
+    """
+    for name in os.listdir(output_dir):
```

Comment:
We should parallelize the upload to s3. I did this in the hca cli and saw 5x improvement

Reply:
The upload is relatively small/quick as we're uploading text files (seconds compared to hours). Granted it's worth double-checking the size of the expression files being uploaded. Will do a quick check on this.

Reply:
Its worth also checking the # of files uploaded. Even if they are small, there tend to be huge wins in parallelization